blockchain/chain.go

```diff
@@ -272,8 +272,47 @@ type BlockChain struct {
 	// so that the memory may be restored by the garbage collector if
 	// it is unlikely to be referenced in the future.
 	pruner *chainPruner
+
+	// The following maps are various caches for the stake version/voting
+	// system.  The goal of these is to reduce disk access to load blocks
+	// from disk.  Measurements indicate that it is slightly more expensive
+	// so setup the cache (<10%) vs doing a straight chain walk.  Every
+	// other subsequent call is >10x faster.  For example:
+	// stakeversion_test.go:930: running: shallow no cache with 20160 nodes
+	// stakeversion_test.go:1000:	run time (setup 0) 712.273µs
+	// stakeversion_test.go:1000:	run time (run 1) 576.898µs
+	// stakeversion_test.go:1000:	run time (run 2) 571.683µs
+	// stakeversion_test.go:1000:	run time (run 3) 571.424µs
+	// stakeversion_test.go:1000:	run time (run 4) 570.918µs
+	// stakeversion_test.go:930: running: shallow cache with 20160 nodes
+	// stakeversion_test.go:1000:	run time (setup 0) 775.044µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 1) 74.316µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 2) 57.283µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 3) 57.235µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 4) 57.068µs memory cost: 116
+	// stakeversion_test.go:930: running: deep no cache with 201600 nodes
+	// stakeversion_test.go:1000:	run time (setup 0) 700.987µs
+	// stakeversion_test.go:1000:	run time (run 1) 571.4µs
+	// stakeversion_test.go:1000:	run time (run 2) 569.362µs
+	// stakeversion_test.go:1000:	run time (run 3) 569.154µs
+	// stakeversion_test.go:1000:	run time (run 4) 569.319µs
+	// stakeversion_test.go:930: running: deep cache with 201600 nodes
+	// stakeversion_test.go:1000:	run time (setup 0) 723.686µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 1) 57.565µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 2) 55.698µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 3) 55.478µs memory cost: 116
+	// stakeversion_test.go:1000:	run time (run 4) 56.046µs memory cost: 116
+	isVoterMajorityVersionCache   map[[stakeMajorityCacheKeySize]byte]bool
+	isStakeMajorityVersionCache   map[[stakeMajorityCacheKeySize]byte]bool
+	calcPriorStakeVersionCache    map[[chainhash.HashSize]byte]uint32
+	calcVoterVersionIntervalCache map[[chainhash.HashSize]byte]uint32
+	calcStakeVersionCache         map[[chainhash.HashSize]byte]uint32
 }
 
+const (
+	stakeMajorityCacheKeySize = 4 + chainhash.HashSize
```

Comment:
Why is this 4?   (Yes, I know from our discussions it's the height, but please comment this).